### DISCUSSION TITLE
ESP32-D0WD-V3 (revision 3) - I2sMode.AdcBuiltIn

### DISCUSSION THREAD

**Discussion by MarP13** [nanoframework/Home#1218] [Q&A]
Hello! First I wanna express my gratitude for this project. Running C#/.NET on microcontrollers has been my wet dream for years and I was very sad to see official microFw die before the price for devices become reasonable. 
It's great to see the resurection with nanoFramework, especially on my beloved ESP32!

To my actual question:
I want to read analog data at high pace. I haven't found option how to set the sampling rate for ADC. 
Is that possible with nano framework?

I found that I2S could be used for this purpose, but I struggle to configure it for ADC.  When I use Mode = AdcBuiltIn program gets stuck on reading into buffer. 

[Espressif's doc for i2s](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/i2s.html#i2s-communication-mode) states, that Adc mode is available on I2S0 bus. Does that equal to nanoframework's busId = 1? I2sConnectionSettings will fail with Id 0 and there are no definitions in DeviceFunction for I2S0_*.

I looked briefly into [I2S implementation](https://github.com/nanoframework/nf-interpreter/blob/017f46d2ea9c9cba229ddba74061d3eb17e43ebe/targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp) for clues, but it's to complex for my lame C++ skill.

Thanks a lot for your help.

```C#
using System;
using System.Device.I2s;
using System.Diagnostics;
using System.Threading;
using nanoFramework.Hardware.Esp32;

namespace NanoFrameworkTest
{
    public class Program
    {
        public static void Main()
        {
           Debug.WriteLine("Hello from I2S!");
            Configuration.SetPinFunction(0, DeviceFunction.I2S1_WS);
            Configuration.SetPinFunction(34, DeviceFunction.I2S1_MDATA_IN);

            I2sDevice i2s = new(new I2sConnectionSettings(1)
            {
                BitsPerSample = I2sBitsPerSample.Bit16,
                ChannelFormat = I2sChannelFormat.AllRight,
                Mode = I2sMode.Master | I2sMode.Rx | I2sMode.AdcBuiltIn,
                CommunicationFormat = I2sCommunicationFormat.I2S,
                SampleRate = 44_100
            });

            SpanByte buff = new byte[1024];
            i2s.Read(buff);
            var x = buff.ToArray();

            Thread.Sleep(Timeout.Infinite);

        }

    }
}
```

**Comment by josesimoes** [comment]
@MarP13 on behalf of the project: thank you for your kind words of appreciation!
Happy to know that you're having fun with .NET nanoFramework.

To your question: sample rate is not exposed in C# API therefore you can't set it using the ADC API.

Regarding I2S, yes, that could be a good alternative as it allows all sorts of configurations. See [here](https://docs.nanoframework.net/api/System.Device.I2s.I2sDevice.html).

Bus indexes in .NET are 1 "based", so you have to refer to it as I2S1 (take a look at the `SetPinFunction` call 😉  you see there `DeviceFunction.I2S1....`

**Comment by MarP13** [comment]
Thank you for swift answer.I did check the docs and tried different configs (slavemode, different buffer sizes, channelconfig and more), but AdcMode keeps failing. The only way I am able to readout data is PDM mode. But I would like to get raw adc data since I won't be receiving sound. On Jan 26, 2023 00:46, José Simões ***@***.***> wrote:

@MarP13 on behalf of the project: thank you for your kind words of appreciation!

Happy to know that you're having fun with .NET nanoFramework.

To your question: sample rate is not exposed in C# API therefore you can't set it using the ADC API.

Regarding I2S, yes, that could be a good alternative as it allows all sorts of configurations. See here.

Bus indexes in .NET are 1 "based", so you have to refer to it as I2S1 (take a look at the SetPinFunction call 😉  you see there DeviceFunction.I2S1....

**Reply by josesimoes** [reply]
@BWalti any thoughts on this?

**Comment by BWalti** [comment]
@MarP13 first, about I2S0: on [this code line](https://github.com/nanoframework/nf-interpreter/blob/017f46d2ea9c9cba229ddba74061d3eb17e43ebe/targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp#L265) you can see that the C# input "BusId" will be reduced by one -> therefore a BusId=1 equals the 0 bus natively.

About the SampleRate: I'd say it is [exposed](https://github.com/nanoframework/System.Device.I2s/blob/main/System.Device.I2s/I2sConnectionSettings.cs#L125) and [consumed](https://github.com/nanoframework/nf-interpreter/blob/017f46d2ea9c9cba229ddba74061d3eb17e43ebe/targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp#L175)..

I've had however problems in understanding what I was doing while implementing I2S natively.. because all these PCM | ADM modes etc.. I really struggled to wrap my head around that stuff.. So native implementation is a bit of: "copy & paste" from some sample source codes and adapting that to nanoFramework.

If you find a "complete" native implementation for ADC and could show me, I could try to figure out, where the differences might be.
I'm not sure if I'd be able to test it though - if you'd have a suggestion / link for that, I would highly appreciate that :)

**Reply by MarP13** [reply]
Thanks for the answer, BWalti.

Oh, I missed that one, thanks for confirmation. 

My original comment was targeting sampling rate on the regular ADC channel, not I2S. If that would be available and gave me similar performance, I would go with that as with simpler solution. 

Regarding the I2S , my understanding of I2S is for sure worse then yours and I haven't found the actual implementation.
Nonetheless, I noticed there are additional methods called to setup I2S together with ADC. Please see example [here](https://github.com/epozzobon/esp32scope),  docs [here](https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32/api-reference/peripherals/adc.html) and discussion/example [here](https://www.esp32.com/viewtopic.php?t=2346#p37378).

EDIT: The file 'dig_i2s_adc.c' [example](https://www.esp32.com/download/file.php?id=2721&sid=8dc7fe313dfa3222cddb53d9f5e508b8) from last link looks a bit as the actual implementation of that ADC mode.

**Reply by BWalti** [reply]
uh yeah, that looks completely different (as in different native calls to "prepare, start and stop" the process).
I assume we could build this into our native code with some "if/else" structure, but I'd have to try it out.

**Reply by MarP13** [reply]
I did little more research yeasterday. Here is a nice example of the use:
https://github.com/espressif/arduino-esp32/blob/master/libraries/ESP32/examples/I2S/HiFreq_ADC/HiFreq_ADC.ino

I set up the nanoframework dev container for ESP, if am not mistaken, full espressif's I2S driver is there right? 
So we should be good just by making methods "i2s_set_adc_mode", "i2s_adc_enable" and maybe "adc1_config_channel_atten" available.

Or even better, calling them from the implementation based on configured I2S ADC mode and PIN function. 
That would not even affect the contract of the library.  I guess, this is what you meant by if/else structure, right?

**Reply by BWalti** [reply]
oh yeah - that looks much closer to the actual implementation and would probably be an easier way to go. Do you want to try it out? As you already have the DevContainer and everything working? ;)

**Reply by MarP13** [reply]
I have a prototype working. :-)  This is 8khz signal sampled at 1MPS rate. 
![ScopeTest](https://user-images.githubusercontent.com/88884514/215354062-e81d31f9-bad9-4101-b837-bb04865d67a8.jpg)


Once I found more time for this, I'll study and think off all ESP32 variants and test the other I2S modes/samples to mare sure I haven't broke anything. Nonetheless, I might not have required HW for all I2S (audio) usecases. I'll also check you contribution rules and post the PR then.

**Reply by BWalti** [reply]
Looks great! I am looking forward to your PR :) If I'm not "linked" with the PR automatically, could you please @ me in a comment? Would be very interested to see how you needed to adapt the code for that :)

**Reply by josesimoes** [reply]
@MarP13 that sounds great!! Looking forward to your PR.
If your changes are touching a lot of code and/or several targets, don't put too much focus on making it perfect. Better start a draft PR so you can get early comments and help from review. We'll be more than happy to provide advice on all that.

**Reply by MarP13** [reply]
Thanks for the support. I struggle to find a time lately. But I created an issue #1222 already and will create PR soon hopefully. 
Let's continue disscussion on the issue & PR.
